diff --git a/auto/lib/libgd/conf b/auto/lib/libgd/conf
index 67863976..8ccf0e89 100644
--- a/auto/lib/libgd/conf
+++ b/auto/lib/libgd/conf
@@ -14,6 +14,20 @@
     . auto/feature
 
 
+if [ -e "$MINGW_PREFIX/lib" -a $ngx_found = no ]; then
+    # MinGW port
+
+    ngx_feature="GD library in $MINGW_PREFIX/"
+    ngx_feature_path="$MINGW_PREFIX/include"
+    ngx_feature_libs="$MINGW_PREFIX/lib/libgd.a"
+
+    if [ $NGX_RPATH = YES ]; then
+        ngx_feature_libs="-R$MINGW_PREFIX/lib $ngx_feature_libs"
+    fi
+
+    . auto/feature
+fi
+
 if [ $ngx_found = no ]; then
 
     # FreeBSD port
diff --git a/auto/lib/libxslt/conf b/auto/lib/libxslt/conf
index 3063ac7..158c662 100644
--- a/auto/lib/libxslt/conf
+++ b/auto/lib/libxslt/conf
@@ -22,6 +22,20 @@
     . auto/feature
 
 
+if [ -e "$MINGW_PREFIX/lib" -a $ngx_found = no ]; then
+    # MinGW port
+
+    ngx_feature="libxslt in $MINGW_PREFIX/"
+    ngx_feature_path="$MINGW_PREFIX/include/libxml2 $MINGW_PREFIX/include"
+    ngx_feature_libs="$MINGW_PREFIX/lib/libxml2.a $MINGW_PREFIX/lib/libxslt.a $MINGW_PREFIX/lib/libiconv.a $MINGW_PREFIX/lib/liblzma.a -lws2_32 $MINGW_PREFIX/lib/libz.a"
+
+    if [ $NGX_RPATH = YES ]; then
+        ngx_feature_libs="-R$MINGW_PREFIX/lib $ngx_feature_libs"
+    fi
+
+    . auto/feature
+fi
+
 if [ $ngx_found = no ]; then
 
     # FreeBSD port
@@ -105,6 +119,20 @@ fi
     ngx_feature_test="exsltRegisterAll();"
     . auto/feature
 
+if [ -e "$MINGW_PREFIX/lib" -a $ngx_found = no ]; then
+    # MinGW port
+
+    ngx_feature="libexslt in $MINGW_PREFIX/"
+    ngx_feature_path="$MINGW_PREFIX/include/libxml2 $MINGW_PREFIX/include"
+    ngx_feature_libs="$MINGW_PREFIX/lib/libexslt.a $MINGW_PREFIX/lib/libxml2.a $MINGW_PREFIX/lib/libxslt.a $MINGW_PREFIX/lib/libiconv.a $MINGW_PREFIX/lib/liblzma.a -lws2_32 $MINGW_PREFIX/lib/libz.a"
+
+    if [ $NGX_RPATH = YES ]; then
+        ngx_feature_libs="-R$MINGW_PREFIX/lib $ngx_feature_libs"
+    fi
+
+    . auto/feature
+fi
+
 if [ $ngx_found = no ]; then
 
     # FreeBSD port
@@ -157,9 +185,10 @@ fi
 
 
 if [ $ngx_found = yes ]; then
+    ngx_feature_libs="${ngx_feature_libs/$NGX_LIB_LIBXSLT/}"
     if [ $USE_LIBXSLT = YES ]; then
-        CORE_LIBS="$CORE_LIBS -lexslt"
+        CORE_LIBS="$CORE_LIBS $ngx_feature_libs"
     fi
 
-    NGX_LIB_LIBXSLT="$NGX_LIB_LIBXSLT -lexslt"
+    NGX_LIB_LIBXSLT="$NGX_LIB_LIBXSLT $ngx_feature_libs"
 fi
diff --git a/auto/lib/openssl/conf b/auto/lib/openssl/conf
index 4fb52df..6b3d1b8 100644
--- a/auto/lib/openssl/conf
+++ b/auto/lib/openssl/conf
@@ -3,7 +3,15 @@
 # Copyright (C) Nginx, Inc.
 
 
-if [ $OPENSSL != NONE ]; then
+if [ -e "$MINGW_PREFIX/lib/libssl.a" -a $OPENSSL = YES ]; then
+    have=NGX_OPENSSL . auto/have
+    have=NGX_SSL . auto/have
+
+    CORE_INCS="$CORE_INCS $MINGW_PREFIX/include"
+    CORE_DEPS="$CORE_DEPS $MINGW_PREFIX/include/openssl/ssl.h"
+    CORE_LIBS="$CORE_LIBS $MINGW_PREFIX/lib/libssl.a $MINGW_PREFIX/lib/libcrypto.a $NGX_LIBDL $NGX_LIBPTHREAD -lgdi32 -lcrypt32 -lws2_32"
+
+elif [ $OPENSSL != NONE ]; then
 
     case "$CC" in
 
diff --git a/auto/lib/pcre/conf b/auto/lib/pcre/conf
index 20c1caf..8b99a64 100644
--- a/auto/lib/pcre/conf
+++ b/auto/lib/pcre/conf
@@ -3,7 +3,18 @@
 # Copyright (C) Nginx, Inc.
 
 
-if [ $PCRE != NONE ]; then
+if [ -e "$MINGW_PREFIX/include/pcre2.h" -a $PCRE = YES ]; then
+    PCRE_LIBRARY=PCRE2
+
+    have=NGX_PCRE . auto/have
+    have=NGX_PCRE2 . auto/have
+    have=PCRE2_STATIC . auto/have
+
+    CORE_INCS="$CORE_INCS $MINGW_PREFIX/include/"
+    CORE_DEPS="$CORE_DEPS $MINGW_PREFIX/include/pcre2.h"
+    CORE_LIBS="$CORE_LIBS $MINGW_PREFIX/lib/libpcre2-8.a"
+
+elif [ $PCRE != NONE ]; then
 
     if [ -f $PCRE/src/pcre2.h.generic ]; then
 
diff --git a/auto/lib/zlib/conf b/auto/lib/zlib/conf
index 239592e..4e381b8 100644
--- a/auto/lib/zlib/conf
+++ b/auto/lib/zlib/conf
@@ -3,7 +3,13 @@
 # Copyright (C) Nginx, Inc.
 
 
-if [ $ZLIB != NONE ]; then
+if [ -e "$MINGW_PREFIX/lib/libz.a" -a $ZLIB = YES ]; then
+    have=NGX_ZLIB . auto/have
+
+    CORE_INCS="$CORE_INCS $MINGW_PREFIX/include"
+    CORE_LIBS="$CORE_LIBS $MINGW_PREFIX/lib/libz.a"
+
+elif [ $ZLIB != NONE ]; then
     CORE_INCS="$CORE_INCS $ZLIB"
 
     case "$NGX_CC_NAME" in
